[JSC] JSON.parse should use simple recursion until it hits soft-recursion threshold

Constellation opened the pull request:
#### 3a9af61361977e1f775f847136d0cf239182c526
<pre>
[JSC] JSON.parse should use simple recursion until it hits soft-recursion threshold
<a href="https://bugs.webkit.org/show_bug.cgi?id=256349">https://bugs.webkit.org/show_bug.cgi?id=256349</a>
rdar://108930906

Reviewed by Alexey Shvayka and Alexey Proskuryakov.

Let&apos;s do super simple JSON.parse until it hits soft-recursion threshold. Our existing code
handles JSON.parse without recursion, this is great, but it is very complicated. And most of
JSON can be parsed with recursion with the current stack capacity. Let&apos;s start with super
simple recursion path, and check soft stack limit carefully. And if we hit, then stop the
recursion and fallback to our generic code. Good thing is that JSON&apos;s leaf object creation
is isolated, so we can start this generic code at any time. We can repeatedly enter and exit
this.

                                                         ToT                     Patched

    vanilla-es2015-babel-webpack-todomvc-json-parse
                                                   63.0883+-0.1159     ^     58.6921+-0.1383        ^ definitely 1.0749x faster
    flight-todomvc-json-parse                      28.3211+-0.0977     ^     25.8927+-0.1618        ^ definitely 1.0938x faster
    vanilla-es2015-todomvc-json-parse              63.1916+-0.2224     ^     58.8631+-0.2339        ^ definitely 1.0735x faster
    vanilla-todomvc-json-parse                     46.7609+-0.1348     ^     43.9075+-0.1779        ^ definitely 1.0650x faster

* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::LiteralParser&lt;CharType&gt;::tryJSONPParse):
(JSC::LiteralParser&lt;CharType&gt;::Lexer::next):
(JSC::LiteralParser&lt;CharType&gt;::parseRecursivelyEntry):
(JSC::LiteralParser&lt;CharType&gt;::parseRecursively):
(JSC::LiteralParser&lt;CharType&gt;::parse):
* Source/JavaScriptCore/runtime/LiteralParser.h:
(JSC::LiteralParser::tryLiteralParse):

Canonical link: <a href="https://commits.webkit.org/263744@main">https://commits.webkit.org/263744@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/965e60c717c0bc3645d7e0b0f1293dad52339bc0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5647 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5800 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5990 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7198 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5632 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5646 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5772 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7759 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5751 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5802 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5058 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7240 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/5753 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3278 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5084 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/12278 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/4696 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5147 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5160 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6988 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/5230 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5595 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4577 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/5757 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5043 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5014 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1409 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9154 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5919 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/646 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5404 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1528 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov reviewed (changes requested):
With this PR, we lose basically all automated testing for the complex code path, which seems not OK.

I think that WebKitTestRunner has to disable the fast path by default, and have a test or two to explicitly verify that it works. And probably something for JSC tests too.

Constellation committed:
[JSC] JSON.parse should use simple recursion until it hits soft-recursion threshold
https://bugs.webkit.org/show_bug.cgi?id=256349
rdar://108930906

Reviewed by Alexey Shvayka and Alexey Proskuryakov.

Let's do super simple JSON.parse until it hits soft-recursion threshold. Our existing code
handles JSON.parse without recursion, this is great, but it is very complicated. And most of
JSON can be parsed with recursion with the current stack capacity. Let's start with super
simple recursion path, and check soft stack limit carefully. And if we hit, then stop the
recursion and fallback to our generic code. Good thing is that JSON's leaf object creation
is isolated, so we can start this generic code at any time. We can repeatedly enter and exit
this.

                                                         ToT                     Patched

    vanilla-es2015-babel-webpack-todomvc-json-parse
                                                   63.0883+-0.1159     ^     58.6921+-0.1383        ^ definitely 1.0749x faster
    flight-todomvc-json-parse                      28.3211+-0.0977     ^     25.8927+-0.1618        ^ definitely 1.0938x faster
    vanilla-es2015-todomvc-json-parse              63.1916+-0.2224     ^     58.8631+-0.2339        ^ definitely 1.0735x faster
    vanilla-todomvc-json-parse                     46.7609+-0.1348     ^     43.9075+-0.1779        ^ definitely 1.0650x faster

* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::LiteralParser<CharType>::tryJSONPParse):
(JSC::LiteralParser<CharType>::Lexer::next):
(JSC::LiteralParser<CharType>::parseRecursivelyEntry):
(JSC::LiteralParser<CharType>::parseRecursively):
(JSC::LiteralParser<CharType>::parse):
* Source/JavaScriptCore/runtime/LiteralParser.h:
(JSC::LiteralParser::tryLiteralParse):

Canonical link: https://commits.webkit.org/263744@main
